executor/insert_common.go

```diff
@@ -795,6 +805,66 @@ func getAutoRecordID(d types.Datum, target *types.FieldType, isInsert bool) (int
 	return recordID, nil
 }
 
+func (e *InsertValues) adjustAutoRandomDatum(ctx context.Context, d types.Datum, hasValue bool, c *table.Column) (types.Datum, error) {
+	if !hasValue || d.IsNull() {
+		_, err := e.ctx.Txn(true)
+		if err != nil {
+			return types.Datum{}, errors.Trace(err)
+		}
+		autoRandomID, err := e.allocAutoRandomID(&c.FieldType)
+		if err != nil {
+			return types.Datum{}, err
+		}
+		d.SetAutoID(autoRandomID, c.Flag)
+	} else {
+		recordID, err := getAutoRecordID(d, &c.FieldType, true)
+		if err != nil {
+			return types.Datum{}, err
+		}
+		err = e.rebaseAutoRandomID(recordID, &c.FieldType)
+		if err != nil {
+			return types.Datum{}, err
+		}
+		d.SetAutoID(recordID, c.Flag)
+	}
+
+	casted, err := table.CastValue(e.ctx, d, c.ToInfo())
+	if err != nil {
+		return types.Datum{}, err
+	}
+	return casted, nil
+}
+
+// allocAutoRandomID allocates a random id for primary key column. It assumes the tableInfo.AutoRandomBits is greater than zero.
+func (e *InsertValues) allocAutoRandomID(fieldType *types.FieldType) (int64, error) {
+	alloc := e.Table.Allocator(e.ctx, autoid.AutoRandomType)
+	tableInfo := e.Table.Meta()
+	_, autoRandomID, err := alloc.Alloc(tableInfo.ID, 1)
+	if err != nil {
+		return 0, err
+	}
+
+	typeBitsLength := uint64(mysql.DefaultLengthOfMysqlTypes[fieldType.Tp] * 8)
+	if tables.OverflowShardBits(autoRandomID, tableInfo.AutoRandomBits, typeBitsLength) {
+		return 0, autoid.ErrAutoincReadFailed
+	}
+	shard := tables.CalcShard(tableInfo.AutoRandomBits, e.ctx.GetSessionVars().TxnCtx.StartTS, typeBitsLength)
+	autoRandomID |= shard
+	return autoRandomID, nil
+}
+
+func (e *InsertValues) rebaseAutoRandomID(recordID int64, fieldType *types.FieldType) error {
+	alloc := e.Table.Allocator(e.ctx, autoid.AutoRandomType)
+	tableInfo := e.Table.Meta()
+
+	typeBitsLength := uint64(mysql.DefaultLengthOfMysqlTypes[fieldType.Tp] * 8)
+	signBit := uint64(1)
+	mask := (1 << (typeBitsLength - tableInfo.AutoRandomBits - signBit)) - 1
+	autoRandomID := int64(mask) & recordID
+
+	return alloc.Rebase(tableInfo.ID, autoRandomID, true)
```

Comment:
Should rebase in update when the `auto_random` column is changed?
```sql
mysql>create table t2 (a bigint key auto_random(1),b int);
Query OK, 0 rows affected
Time: 0.006s
mysql>insert into t2 values(null, 11);
Query OK, 1 row affected
Time: 0.003s
mysql>insert into t2 values(null, 1);
Query OK, 1 row affected
Time: 0.001s
mysql>select * from t2
+---------------------+----+
| a                   | b  |
+---------------------+----+
| 2                   | 1  |
| 4611686018427387905 | 11 |
+---------------------+----+
2 rows in set
Time: 0.007s
mysql>update t2 set a=a+1;
Query OK, 2 rows affected
Time: 0.001s
mysql>insert into t2 values(null, 1);
(1062, u"Duplicate entry '3' for key 'PRIMARY'")
```
Just like the auto_increment.  seach the comment `// Rebase auto increment id if the field is changed.`

Reply:
The reason for `duplicate entry` error is rebase, see:

https://github.com/pingcap/tidb/pull/13127/files#diff-0f31e77a0899889b749ff90c69acaaf6R824

The risk of this feature is that users can modify the value of `auto_random` column but should be careful, the rebase behavior may be out of expectation:

> Modification on the value of auto_random is strongly not recommended.

Reply:
Updated, PTAL